### PR TITLE
Add rake task to send mfa notification

### PR DESCRIPTION
**Pull request by sonalkr132**
Open to suggestions on making this more considerate and persuasive.
Related: #2057

Screenshot from 2019-08-26 23-53-05

### CODE REVIEW

File: lib/tasks/mfa_notification.rake [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+  puts "Backtrace:\n\t#{e.backtrace.join("\n\t")}"
+  false
+end
+
+namespace :mfa_notification do
+  desc "Send email notification to all users about MFA"
+  task send: :environment do
+    i = 0
+    total = User.count
+    User.find_each do |user|
```
Comment (eliotsykes) [inline lib/tasks/mfa_notification.rake]: Consider:

- skipping users who already have MFA enabled for both UI+API.
- limit to sending in small batches of users over a few days/weeks, allowing time for feedback on the email content from the early recipients.

File: app/views/mailer/mfa_notification.html.erb [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+<table width="100%" border="0" cellspacing="0" cellpadding="0" bgcolor="#ffffff">
+  <tr>
+    <td class="content-spacing" style="font-size:0pt; line-height:0pt; text-align:left" width="20"></td>
+    <td>
+      <table width="100%" border="0" cellspacing="0" cellpadding="0" class="spacer" style="font-size:0pt; line-height:0pt; text-align:center; width:100%; min-width:100%"><tr><td height="35" class="spacer" style="font-size:0pt; line-height:0pt; text-align:center; width:100%; min-width:100%">&nbsp;</td></tr></table>
+
+      <table width="100%" border="0" cellspacing="0" cellpadding="0" class="spacer" style="font-size:0pt; line-height:0pt; text-align:center; width:100%; min-width:100%"><tr><td height="40" class="spacer" style="font-size:0pt; line-height:0pt; text-align:center; width:100%; min-width:100%">&nbsp;</td></tr></table>
+
+      <div class="h3-1-center" style="color:#1e1e1e; font-family:Georgia, serif; min-width:auto !important; font-size:20px; line-height:26px; text-align:center">
+        We care about keeping your RubyGem.org account secure and recommend enabling multifactor authentication for <b>both UI and API</b>.
```
Comment (eliotsykes) [inline app/views/mailer/mfa_notification.html.erb]: Typo on RubyGem.org -> RubyGems.org

File: app/views/mailer/mfa_notification.html.erb [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        Setting up MFA only takes a few minutes and goes a long way towards protecting your account from any malicious actor.
+        <br/>
```
Comment (eliotsykes) [inline app/views/mailer/mfa_notification.html.erb]: Mentioning this helps protect the Ruby ecosystem might persuade a few more users to enable MFA, e.g.

> Setting up MFA only takes a few minutes and goes a long way towards protecting your account and the Ruby ecosystem.

File: app/views/mailer/mfa_notification.html.erb [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+            <table width="210" border="0" cellspacing="0" cellpadding="0">
+              <tr>
+                <td align="center" bgcolor="#e9573f">
+                  <table border="0" cellspacing="0" cellpadding="0">
+                    <tr>
+                      <td class="img" style="font-size:0pt; line-height:0pt; text-align:left" width="15"><table width="100%" border="0" cellspacing="0" cellpadding="0" class="spacer" style="font-size:0pt; line-height:0pt; text-align:center; width:100%; min-width:100%"><tr><td height="50" class="spacer" style="font-size:0pt; line-height:0pt; text-align:center; width:100%; min-width:100%">&nbsp;</td></tr></table>
+</td>
+                      <td bgcolor="#e9573f">
+                        <div class="text-btn" style="color:#ffffff; font-family:Arial, sans-serif; min-width:auto !important; font-size:16px; line-height:20px; text-align:center">
+                          <a href=<%= new_multifactor_auth_url %> target="_blank" class="link-white" style="color:#ffffff; text-decoration:none"><span class="link-white" style="color:#ffffff; text-decoration:none">ENABLE MFA</span></a>
```
Comment (eliotsykes) [inline app/views/mailer/mfa_notification.html.erb]: A "login/enable mfa" link in an email might smell phishy to some users. Is this a problem?

File: lib/tasks/mfa_notification.rake (hunk 1 above)
Comment (sonalkr132) [reply lib/tasks/mfa_notification.rake]: Thanks. We will send mail to users with login in 2019 first and possible after two weeks for rest.

File: app/views/mailer/mfa_notification.html.erb (hunk 4 above)
Comment (sonalkr132) [reply app/views/mailer/mfa_notification.html.erb]: Kind of. I don't see any alternative tho. This is better than making them search for the enable link.

File: lib/tasks/mfa_notification.rake [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+
+  return false if mx_resolver.getresources(domain, Resolv::DNS::Resource::IN::MX).empty?
+  true
+rescue StandardError => e
+  puts "Error during processing: #{$ERROR_INFO}"
+  puts "Backtrace:\n\t#{e.backtrace.join("\n\t")}"
+  false
+end
+
+# login in 2019: rake mfa_notification:send
```
Comment (eliotsykes) [inline lib/tasks/mfa_notification.rake]: > Thanks. We will send mail to users with login in 2019 first and possible after two weeks for rest.

Sounds good. Is there potential to miss a few users in the 2nd batch. E.g.

1. User A last signed-in in 2018.
2. Send 1st batch of emails with `rake mfa_notification:send`. User A doesn't receive this email.
3. User A signs in, their updated_at is now in 2019.
4. Send 2nd batch of emails with `rake mfa_notification:send[2009-01-01,2018-12-31]`. User A doesn't receive email again.

Comment (sonalkr132) [reply lib/tasks/mfa_notification.rake]: Thanks for pointing out. We will use `rake mfa_notification:send[2019-01-01,2019-06-01]` at first and following for next iteration:
```
rake mfa_notification:send[2009-01-01,2018-12-31]
rake mfa_notification:send[2019-06-02]
```
Some users will get notified twice, which shouldn't be a huge deal.

File: app/views/mailer/mfa_notification.html.erb (hunk 4 above)
Comment (eliotsykes) [reply app/views/mailer/mfa_notification.html.erb]: One thing might be to post an accompanying article on the RubyGems.org blog mentioning the recent incidents and why MFA is a good idea. Plus include a link to the blog post in the email.

**Pull request merged**